src/zino/trapobservers/link_traps.py

```diff
@@ -0,0 +1,181 @@
+"""This module implements link trap handling"""
+
+import logging
+import re
+from datetime import datetime, timedelta
+from typing import Optional, Tuple
+
+import zino.time
+from zino.statemodels import DeviceState, InterfaceState, Port, PortStateEvent
+from zino.trapd import TrapMessage, TrapObserver
+
+_logger = logging.getLogger(__name__)
+
+
+class LinkTrapObserver(TrapObserver):
+    WANTED_TRAPS = {
+        ("IF-MIB", "linkUp"),
+        ("IF-MIB", "linkDown"),
+    }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._last_same_trap: dict[Tuple[str, int], datetime] = {}
+
+    def handle_trap(self, trap: TrapMessage) -> Optional[bool]:
+        _logger.debug("%s: %s (vars: %s)", trap.agent.device.name, trap.name, ", ".join(trap.variables))
+
+        if "ifIndex" in trap.variables:
+            ifindex = trap.variables.get("ifIndex").value
+        else:
+            ifindex = -1
+        port = trap.agent.device.ports.get(ifindex) if ifindex > 0 else None
+
+        # TODO: The trap *might* contain an ifDescr value.  If present, Zino uses that for trap processing.
+        #  Otherwise, it fetches ifDescr from its own state and uses that for trap processing.  Either way,
+        #  there seems to be some redundancy. We should document why, or change the behavior in Zino 2
+
+        # The legacy Zino also looks for `locIfDescr` in the trap at this point, but this is an ancient
+        # Cisco-specific variable we no longer see.  It might have been replaced by `cieIfOperStatusCause`,
+        # but we have no more Cisco devices to test on, so this has been deliberately left out.
+
+        is_up = trap.name == "linkUp"
+        self.handle_link_transition(trap.agent.device, port, is_up)
```

Comment:
Port is set to `None` here if `ifindex > 0` and then put as argument to `self.handle_link_transition` which does not accept `None` as a value for `port`. The `port` value is then further used in other functions that only accept the `Port` type. 

I imagine the correct thing to do is update these typehints to `Optional[Port]`

Reply:
I followed the original implementation a bit too closely here, methinks.  

It's not really clear *what* does happen in the original when the trap doesn't contain an `ifIndex` value, except that it's set internally to `-1` and then used for lookups.  But of course, it won't find such an interface in its internal data structures.

I'm thinking the best strategy here is to drop the trap immediately, since it's of no use.

Reply:
Added a commit to this effect